torchao/quantization/quant_api.py

```diff
@@ -510,6 +512,41 @@ def int8_dynamic_activation_int4_weight(group_size=32, mapping_type=MappingType.
     return _get_linear_subclass_inserter(apply_int8_dynamic_activation_int4_weight_quant, group_size=group_size, mapping_type=mapping_type)
 
 
+def apply_int8_dynamic_activation_int4_weight_quant_cutlass(weight):
+    """This is defined here instead of local function to support serialization
+    """
+    # weight settings
+    mapping_type = MappingType.SYMMETRIC
+    block_size = (1, weight.shape[-1])
+    target_dtype = torch.int8
+    eps = torch.finfo(torch.float32).eps
+    quant_min = -8
+    quant_max = 7
+
+    # input settings
+    input_quant_func = _int8_symm_per_token_reduced_range_quant_cutlass
+
+    weight = to_affine_quantized_intx(weight, mapping_type, block_size, target_dtype, quant_min, quant_max, eps)
+    weight = to_linear_activation_quantized(weight, input_quant_func)
+
+    # FIXME: this should be done by to_affine_quantized_intx, also the
+    # dtype of quantized tensor maybe should be set to quint4x2, and
+    # then corresponding changes made in
+    # _linear_int8_act_int4_weight_cutlass_check and for the check in
+    # the CUTLASS kernel!!!
+    weight.original_weight_tensor.layout_tensor.int_data = (
+        (weight.original_weight_tensor.layout_tensor.int_data[:, 1::2] & 0xF) << 4
+    ) | (weight.original_weight_tensor.layout_tensor.int_data[:, 0::2] & 0xF)
+
+    return weight
+
+def int8_dynamic_activation_int4_weight_cutlass():
+    """Applies int8 dynamic per token asymmetric activation quantization and int4 per group weight symmetric quantization to linear
+    This is used to produce a model for CUTLASS-based W4A8 kernel
+    """
+    return _get_linear_subclass_inserter(apply_int8_dynamic_activation_int4_weight_quant_cutlass)
+
+
 def int4_weight_only(group_size=128, layout_type=TensorCoreTiledLayoutType(inner_k_tiles=8), use_hqq=False):
```

Comment:
Unrelated comment, what is this `use_hqq`? @jerryzh168 do you know?

Reply:
yeah this means use `hqq` algorithm to choose qparams and quantize the weight, since it is reusing the tinygemm kernel, we just added this as a separate option here